modules/admission-webhook-create.adoc

```diff
@@ -0,0 +1,414 @@
+// Module included in the following assemblies:
+//
+// * architecture/admission-controllers.adoc
+
+[id="admission-webhook-create_{context}"]
+= Create an admission webhook
+
+The steps below illustrate how to build a secure and portable webhook admission server.
+
+In this example, the webhook server is also an aggregated API server. This allows other OpenShift components to communicate with the webhook using internal credentials and facilitates testing using the `oc` command. Deploying an aggregated API server in this way also enables role based access control into the webhook, as well as preventing token information from other API servers being disclosed to the webhook. The API server will call `kube-apiserver` with its fully authenticated service account. The aggregated API server then calls to the webhook using the request header certificate. The webhook verifies incoming client certificates through the published request header aggregation CA.
+
+. Create a local CA key and certificate, plus a webhook server key and certificate which are signed by the local CA (replace `<variables>` accordingly):
++
+----
+mkdir _output/tmp/certs
+----
++
+----
+openssl req -x509 -sha256 -new -nodes -days 365 -newkey rsa:2048 -keyout "_output/tmp/certs/serving-ca.key" -out "_output/tmp/certs/serving-ca.crt" -subj "/C=<country>/ST=<state>/L=<city>/O=<company_name>/OU=<organizational_unit>/CN=ca/emailAddress=<email_address>/"
+----
++
+----
+echo '{"signing":{"default":{"expiry":"43800h","usages":["signing","key encipherment","server auth"]}}}' > "_output/tmp/certs/serving-ca-config.json"
+----
+
+. Create a server key and certificate for the webhook:
++
+----
+cd _output/tmp/certs
+----
++
+----
+echo '{"CN":"server.openshift-namespace-reservation.svc","hosts":[server.openshift-namespace-reservation.svc],"key":{"algo":"rsa","size":2048}}' | cfssl gencert -ca=${ca}.crt -ca-key=serving-ca.key -config=serving-ca-config.json - | cfssljson -bare serving-server.openshift-namespace-reservation.svc
+----
++
+----
+mv "serving-server.openshift-namespace-reservation.svc-key.pem" "serving-server.openshift-namespace-reservation.svc.key"
+----
++
+----
+mv "serving-server.openshift-namespace-reservation.svc.pem" "serving-server.openshift-namespace-reservation.svc.crt"
+----
++
+----
+rm -f "serving-server.openshift-namespace-reservation.svc.csr"
+----
+
+. Make sure the OpenShift configuration context is set accordingly.
++
+----
+oc config current-context
+----
++
+If it is not, then set it from one listed output by:
++
+----
+oc config get-contexts
+----
++
+----
+oc config set-context <context> --user=<cluster_admin>
+----
+
+. Create a YAML configuration file specifying RBAC rules for the aggregated API service:
++
+----
+mkdir yaml_files
+----
++
+----
+vi yaml_files/rbac-list.yaml
```

Comment:
I'd skip suggesting the vi editor and just say "Create a file called `rbac-list.yaml` with the following contents:

Reply:
Done.